[jax2tf] First draft of top_k conversion to tf.

bchetioui opened the pull request:
We are able to use tf.math.top_k in most cases but:
- bfloat16 and complex numbers are unsupported
- tf.math.top_k works properly with args of type int32/64, uint32/64 and
  float32/64. Other flavors of uints/ints and bool can not be used with
  the operation by default, and they are thus promoted to their
  equivalent 32 bit representation, and the result is casted back to the
  original dtype.

A limitation right now is the handling of edge cases: TF and JAX order
np.inf and np.nan differently, resulting in inconsistencies in the case
where these values are present.

bchetioui committed:
[jax2tf] First draft of top_k conversion to tf.

We are able to use tf.math.top_k in most cases but:
- bfloat16 and complex numbers are unsupported
- tf.math.top_k works properly with args of type int32/64, uint32/64 and
  float32/64. Other flavors of uints/ints and bool can not be used with
  the operation by default, and they are thus promoted to their
  equivalent 32 bit representation, and the result is casted back to the
  original dtype.

A limitation right now is the handling of edge cases: TF and JAX order
np.inf and np.nan differently, resulting in inconsistencies in the case
where these values are present.

gnecula commented on jax/experimental/jax2tf/jax2tf.py:
```diff
@@ -1113,6 +1113,29 @@ def func1(*jax_args):
+    if dtype in [jnp.float16]:
+      return jnp.float32
+    if dtype in [jnp.bfloat16]:
+      return tf.bfloat16
```
I am wondering why we return here `tf.bfloat16`, and in all other branches we returns a `jnp` type?

SIben replied:
Addressed in latest commits.

gnecula commented on jax/experimental/jax2tf/jax2tf.py:
```diff
@@ -1113,6 +1113,29 @@ def func1(*jax_args):
+def _top_k(operand: TfVal, k: int) -> Tuple[TfVal, TfVal]:
+  # Some types originally incompatible with tf.math.top_k can be promoted
+  # to a compatible type without loss of precision.
+  def promote_dtype(dtype):
```
It seems that `dtype` here is a TF dtype (result of `operand.dtype`). If that is true, then it would be clearer if we compare `dtype` with `numpy` types, not `jax.numpy` types (e.g., `dtype in [np.bool_, np.uint8, ...]`)

SIben replied:
Addressed in latest commits.

gnecula commented on jax/experimental/jax2tf/tests/primitives_test.py:
```diff
@@ -109,6 +109,24 @@ def test_pad(self, harness: primitive_harness.Harness):
 
+  @primitive_harness.parameterized(primitive_harness.lax_top_k)
+  def test_top_k(self, harness: primitive_harness.Harness):
+    if (harness.params["k"] > harness.params["shape"][-1] or
```
I do not think that we need to test this here. The goal is to test the conversion of primitive usage (lax.top_k_p.bind). The negative `k` will be rejected before we even get to use the primitive.

SIben replied:
I believe that we should keep these tests to make sure `lax.top_k` always fails in potential edge cases. Because the clamping behavior is not always consistent, it is important that we know if that invalid parameters are never silently passed to the function.

gnecula commented on jax/experimental/jax2tf/tests/primitive_harness.py:
```diff
@@ -319,6 +319,40 @@ def parameterized(harness_group: Iterable[Harness],
+          k=k)
+  for dtype in jtu.dtypes.all #[np.float32, np.int32, np.uint32]
+  for shape in [(3,), (5, 3)]
+  for k in [-1, 1, 3, 4]
```
I do not think that we need to include invalid values for `k` here. See comment below.

SIben replied:
See argument below.

gnecula commented on jax/experimental/jax2tf/tests/primitive_harness.py:
```diff
@@ -319,6 +319,40 @@ def parameterized(harness_group: Iterable[Harness],
+          shape=shape,
+          dtype=dtype,
+          k=k)
+  for dtype in jtu.dtypes.all #[np.float32, np.int32, np.uint32]
```
Can we remove the comment at the end of the line?

SIben replied:
Addressed in latest commits.

gnecula commented on jax/experimental/jax2tf/tests/primitive_harness.py:
```diff
@@ -319,6 +319,40 @@ def parameterized(harness_group: Iterable[Harness],
+  for k in [-1, 1, 3, 4]
+  for rng_factory in [jtu.rand_default]
+) + tuple( # stability test
+  Harness(f"stability_inshape={jtu.format_shape_dtype_string(arr.shape, arr.dtype)}_k={k}",
```
I am curious how is stability involved in top_k? Is it about testing cases when we have repeated values?

SIben replied:
`top_k` returns the top k elements as well as their indices within the array; for that reason, it is important to know how the case in which two elements are the same is treated. The [TensorFlow docs](https://www.tensorflow.org/api_docs/python/tf/math/top_k) mentions that "If two elements are equal, the lower-index element appears first.", hence the stability check.

bchetioui committed:
Fix type casting to use tf dtypes.

bchetioui committed:
Added explicit sorted=True parameter to tf.math.top_k.

Note: this has been sanity checked, the results are indeed inconsistent
if we set sorted=False.

bchetioui committed:
Removed comment.

gnecula commented:
I tested this internally before merging.
